File: shorelinecalculator/options.cpp
#include "options.hpp"
#include <stdexcept>
#include <string>

namespace dsas {

Options::Options(const nlohmann::json& json_value) {
  const auto& json_options = json_value.at("options");

  if (json_options.contains("smooth_factor")) {
    smooth_factor = json_options.at("smooth_factor").get<int>();
  }

  if (json_options.contains("edge_distance")) {
    edge_distance = json_options.at("edge_distance").get<int>();
  }

  if (json_options.contains("shoreline_least_factor")) {
    shoreline_least_factor = json_options.at("shoreline_least_factor").get<double>();
  }

  if (json_options.contains("transect_length")) {
    transect_length = json_options.at("transect_length").get<double>();
  }

  if (json_options.contains("transect_spacing")) {
    transect_spacing = json_options.at("transect_spacing").get<double>();
  }

  if (json_options.contains("transect_offset")) {
    transect_offset = json_options.at("transect_offset").get<double>();
  }

  if (json_options.contains("outlier_rate")) {
    outlier_rate = json_options.at("outlier_rate").get<double>();
  }

  if (json_options.contains("thread_num")) {
    thread_num = json_options.at("thread_num").get<size_t>();
  }

  if (json_options.contains("intersection_mode")) {
    std::string mode = json_options.at("intersection_mode").get<std::string>();
    if (mode == "closest") {
      intersection_mode = IntersectionMode::Closest;
    } else if (mode == "farthest") {
      intersection_mode = IntersectionMode::Farthest;
    } else {
      throw std::runtime_error("not a valid intersection mode");
    }
  }

  if (json_options.contains("transect_orientation")) {
    std::string orient = json_options.at("transect_orientation").get<std::string>();
    if (orient == "left") {
      transect_orient = TransectOrientation::Left;
    } else if (orient == "right") {
      transect_orient = TransectOrientation::Right;
    } else if (orient == "mix") {
      transect_orient = TransectOrientation::Mix;
    } else {
      throw std::runtime_error("not a valid orientation");
    }
  }
}

}  // namespace dsas

File: scet/options.py
"""Configuration shared by the image-extraction pipeline and OpenDSAS CLI calls."""
from __future__ import annotations

from dataclasses import dataclass
from enum import Enum
from typing import Any


class IntersectionMode(str, Enum):
    CLOSEST = "closest"
    FARTHEST = "farthest"


class TransectOrientation(str, Enum):
    LEFT = "left"
    RIGHT = "right"
    MIX = "mix"


@dataclass
class Options:
    # image -> shoreline extraction
    edge_distance: int = 100
    shoreline_least_factor: float = 0.5

    # forwarded to `dsas cast` / `dsas cal`
    smooth_factor: int = 1
    transect_length: float = 500
    transect_spacing: float = 30
    intersection_mode: IntersectionMode = IntersectionMode.CLOSEST
    transect_orient: TransectOrientation = TransectOrientation.MIX

    @classmethod
    def from_json(cls, json_value: dict[str, Any]) -> "Options":
        data = json_value.get("options", json_value)
        kwargs: dict[str, Any] = {}
        if "smooth_factor" in data:
            kwargs["smooth_factor"] = int(data["smooth_factor"])
        if "edge_distance" in data:
            kwargs["edge_distance"] = int(data["edge_distance"])
        if "shoreline_least_factor" in data:
            kwargs["shoreline_least_factor"] = float(data["shoreline_least_factor"])
        if "transect_length" in data:
            kwargs["transect_length"] = float(data["transect_length"])
        if "transect_spacing" in data:
            kwargs["transect_spacing"] = float(data["transect_spacing"])
        if "intersection_mode" in data:
            kwargs["intersection_mode"] = IntersectionMode(data["intersection_mode"])
        if "transect_orientation" in data:
            kwargs["transect_orient"] = TransectOrientation(data["transect_orientation"])
        return cls(**kwargs)
